dynamicio/mixins/with_kafka.py

```diff
@@ -89,28 +251,39 @@ def _write_to_kafka(self, df: pd.DataFrame) -> None:
         Args:
             df: A dataframe where each row is a message to be sent to a Kafka Topic.
         """
+        self.populate_cls_attributes()
+
+        if self.__producer is None:
+            self.__producer = self._get_producer(self.sources_config.kafka.kafka_server, **self.options)
+
+        self._send_messages(df=df, topic=self.sources_config.kafka.kafka_topic)
+
+    def populate_cls_attributes(self):
+        """Pop dynamicio options (key_generator, document_transformer, key_serializer, value_serializer) from kafka config options."""
         if self.__key_generator is None:
             self.__key_generator = lambda idx, __: idx  # default key generator uses the dataframe's index
             if self.options.get("key_generator") is not None:
                 self.__key_generator = self.options.pop("key_generator")
-
         if self.__document_transformer is None:
             self.__document_transformer = lambda value: value
             if self.options.get("document_transformer") is not None:
                 self.__document_transformer = self.options.pop("document_transformer")
-
-        if self.__producer is None:
-            self.__producer = self._get_producer(self.sources_config.kafka.kafka_server, **self.options)
-
-        self._send_messages(df=df, topic=self.sources_config.kafka.kafka_topic)
-
-    @utils.allow_options(KafkaProducer.DEFAULT_CONFIG.keys())
-    def _get_producer(self, server: str, **options: MutableMapping[str, Any]) -> KafkaProducer:
+        if self.__key_serializer is None:
```

Comment:
You can just set __key_serializer and __value_serializer to whatever your default functions are, then override if options contain either. Not sure why this indirection exists.

Reply:
Looking into it...

Reply:
Good point but leaving it as is for now as the intended behaviour is properly tested and the refactoring wont change that behaviour. Good point tho.